File: clients/go/delete_offsets_test.go
package volant_test

import (
	"errors"
	"net"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

type deleteOffsetsServerResult struct {
	group   string
	entries []codec.OffsetEntry
	err     error
}

func serveDeleteOffsets(t *testing.T, errorCode uint16, deletedCount uint32) (addr string, got *deleteOffsetsServerResult, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	res := &deleteOffsetsServerResult{}
	done := make(chan struct{})
	go func() {
		defer close(done)
		conn, err := ln.Accept()
		if err != nil {
			res.err = err
			return
		}
		defer conn.Close()
		_ = conn.SetDeadline(time.Now().Add(5 * time.Second))
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			f, rest, err := frame.TryDecode(buf)
			if err != nil {
				res.err = err
				return
			}
			if f == nil {
				n, err := conn.Read(tmp)
				if n > 0 {
					buf = append(buf, tmp[:n]...)
				}
				if err != nil {
					res.err = err
					return
				}
				continue
			}
			buf = append([]byte(nil), rest...)
			if f.Opcode != codec.OpDeleteOffsets {
				res.err = &frame.ProtocolError{Msg: "unexpected opcode"}
				return
			}
			req, err := codec.DecodeDeleteOffsetsRequest(f.Payload)
			if err != nil {
				res.err = err
				return
			}
			res.group = req.GroupID
			res.entries = req.Entries
			payload, err := codec.EncodeDeleteOffsetsResponse(codec.DeleteOffsetsResponse{
				ErrorCode:    errorCode,
				DeletedCount: deletedCount,
			})
			if err != nil {
				res.err = err
				return
			}
			raw, err := frame.Encode(codec.OpDeleteOffsetsResponse, f.CorrelationID, payload)
			if err != nil {
				res.err = err
				return
			}
			if _, err := conn.Write(raw); err != nil {
				res.err = err
				return
			}
		}
	}()
	return ln.Addr().String(), res, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func TestDeleteOffsetsEmptyEntriesEncodedAsCountZero(t *testing.T) {
	addr, got, stop := serveDeleteOffsets(t, 0, 3)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.DeleteOffsets("g", nil)
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.group != "g" || len(got.entries) != 0 {
		t.Fatalf("wire request group=%q entries=%v", got.group, got.entries)
	}
	if out != 3 {
		t.Fatalf("deleted_count %d", out)
	}
}

func TestDeleteOffsetsExplicitEntry(t *testing.T) {
	addr, got, stop := serveDeleteOffsets(t, 0, 1)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.DeleteOffsets("g", []codec.OffsetEntry{{Topic: "events", Partition: 0}})
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.group != "g" || len(got.entries) != 1 || got.entries[0].Topic != "events" || got.entries[0].Partition != 0 {
		t.Fatalf("entries %v", got.entries)
	}
	if out != 1 {
		t.Fatalf("deleted_count %d", out)
	}
}

func TestDeleteOffsetsNonzeroErrorRaises(t *testing.T) {
	addr, _, stop := serveDeleteOffsets(t, 2, 0)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.DeleteOffsets("missing", nil)
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 2 || be.Op != "delete_offsets" {
		t.Fatalf("broker error %+v", be)
	}
}

File: clients/java/src/test/java/io/volant/DeleteOffsetsTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** DeleteOffsets client tests against a scripted TCP broker (no live server). */
class DeleteOffsetsTest {
    @Test
    void emptyEntriesEncodedAsCountZero() throws Exception {
        try (DeleteOffsetsServer srv = DeleteOffsetsServer.ok(3)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                int got = c.deleteOffsets("g");
                assertEquals(3, got);
            }
            assertEquals("g", srv.group.get());
            assertTrue(srv.entries.isEmpty());
        }
    }

    @Test
    void explicitEntryRoundtrip() throws Exception {
        try (DeleteOffsetsServer srv = DeleteOffsetsServer.ok(1)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                int got = c.deleteOffsets(
                        "g", Collections.singletonList(new Codec.OffsetEntry("events", 0)));
                assertEquals(1, got);
            }
            assertEquals("g", srv.group.get());
            assertEquals(1, srv.entries.size());
            assertEquals("events", srv.entries.get(0).topic);
            assertEquals(0, srv.entries.get(0).partition);
        }
    }

    @Test
    void deleteOffsetEncodesOneEntry() throws Exception {
        try (DeleteOffsetsServer srv = DeleteOffsetsServer.ok(1)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                int got = c.deleteOffset("g", "events", 0);
                assertEquals(1, got);
            }
            assertEquals("g", srv.group.get());
            assertEquals(1, srv.entries.size());
            assertEquals("events", srv.entries.get(0).topic);
            assertEquals(0, srv.entries.get(0).partition);
        }
    }

    @Test
    void nonzeroErrorCodeRaises() throws Exception {
        try (DeleteOffsetsServer srv = DeleteOffsetsServer.error(2)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(BrokerException.class, () -> c.deleteOffsets("missing"));
                assertEquals(2, ex.code);
                assertEquals("delete_offsets", ex.op);
            }
        }
    }

    private static final class DeleteOffsetsServer implements AutoCloseable {
        final int port;
        final AtomicReference<String> group = new AtomicReference<>();
        final List<Codec.OffsetEntry> entries = new CopyOnWriteArrayList<>();
        private final int errorCode;
        private final int deletedCount;
        private final ServerSocket listen;
        private final Thread thread;
        private final AtomicReference<Exception> error = new AtomicReference<>();

        static DeleteOffsetsServer ok(int deletedCount) throws IOException {
            return new DeleteOffsetsServer(0, deletedCount);
        }

        static DeleteOffsetsServer error(int code) throws IOException {
            return new DeleteOffsetsServer(code, 0);
        }

        private DeleteOffsetsServer(int errorCode, int deletedCount) throws IOException {
            this.errorCode = errorCode;
            this.deletedCount = deletedCount;
            listen = new ServerSocket(0, 8, InetAddress.getByName("127.0.0.1"));
            listen.setSoTimeout(8_000);
            port = listen.getLocalPort();
            thread = new Thread(this::serve, "volant-delete-offsets");
            thread.setDaemon(true);
            thread.start();
        }

        private void serve() {
            try (Socket conn = listen.accept()) {
                conn.setSoTimeout(5_000);
                InputStream in = conn.getInputStream();
                OutputStream out = conn.getOutputStream();
                byte[] buf = new byte[0];
                while (true) {
                    Frame.Decode d = Frame.tryDecode(buf);
                    if (d.frame == null) {
                        byte[] tmp = new byte[4096];
                        int n = in.read(tmp);
                        if (n < 0) {
                            return;
                        }
                        byte[] next = new byte[buf.length + n];
                        System.arraycopy(buf, 0, next, 0, buf.length);
                        System.arraycopy(tmp, 0, next, buf.length, n);
                        buf = next;
                        continue;
                    }
                    buf = d.rest;
                    if (d.frame.opcode != Codec.OP_DELETE_OFFSETS) {
                        error.set(new ProtocolException("unexpected opcode " + d.frame.opcode));
                        return;
                    }
                    Codec.DeleteOffsetsRequest req = Codec.decodeDeleteOffsetsRequest(d.frame.payload);
                    group.set(req.groupId);
                    entries.clear();
                    entries.addAll(req.entries);
                    byte[] payload = Codec.encodeDeleteOffsetsResponse(
                            new Codec.DeleteOffsetsResponse(errorCode, deletedCount));
                    out.write(Frame.encode(Codec.OP_DELETE_OFFSETS_RESPONSE, d.frame.correlationId, payload));
                    out.flush();
                }
            } catch (Exception e) {
                error.set(e);
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                thread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}
